### ISSUE TITLE
Tab's history gets broken

### ISSUE THREAD

**Issue by Drugoy** [cfinke/Tab-History-Redux#1]
1. Open some page.
2. Middle-click on any link on it: a new tab will get opened.
3. Close the new tab.
4. Hit CTRL+SHIFT+T to undo close tab.
5. Now watch the restored tab's history: it is broken. The list is wrong + you may click "back" and then "forward" and you won't get to the same page: instead, the parental tab's page will be opened.

**Comment by Drugoy** [comment]
WFM on todays Nightly.

**Comment by DavidRegev** [comment]
**Drugoy**: Are you _sure_ it works for you? The latest version came out after the last change, back in May. Using a new nightly should not have made a difference, as the code would be identical.

I have been experiencing the same problem myself for quite some time now, though have only recently managed to narrow it down to this extension. It happens sporadically, and I do not know a deterministic way of reproducing it. It can lead to some real data loss when it happens, and it occurs too often.

Here are the steps that _sometimes_ reproduce this bug most often for me (fewer steps may be needed for others):
1. Set `browser.sessionstore.max_concurrent_tabs` to `0`.
2. Create a few app tabs. (What works for me: Gmail, another Gmail app tab, and Google Reader.)
3. Open and close a bunch of sites, so that there are tabs in the Recently Closed Tabs list.
4. Restart Firefox.
5. Quickly focus every tab so that they all reload at the same time.

**Results**: A lot of the time, there will be a tab that reloads to the wrong page, as if the history leaked from another tab (often a closed one) to that tab.

The bug manifests itself in other ways as well. Sometimes, clicking Back won’t work at all, even though the tab’s history is still visible behind the Back button. Other times, clicking Back will jump far back in the tab history. Yet other times, clicking Back will trigger the results I described above, where a different tab will suddenly take over that tab.

In short, tab histories get corrupted easily.

Could this issue be reopened? I will gladly offer any information or file that maybe needed to get this fixed. Thanks.

**Comment by Drugoy** [comment]
No, the bug came back after that, but as this extension is forgotten and developed by no one, I didn't bother myself re-opening the issue.
But you are wrong about "Using a new nightly should not have made a difference, as the code would be identical." Nightly gets changed everyday, and it doesn't matter that the extension was not updated: instead, Nightly did and that may have affected the situation.
Actually I also experienced that bug by the steps-to-reproduce (STR) you've described, but it's pretty easy to avoid, as the key component here is the phrase "quickly focus every tab". If you do that slowly - it's OK in most cases.
And you are also absolutely right, that the bug is actually way bigger, I just couldn't describe the STR for many other cases when I experienced that bug.

I found out, that Tab Utilities extension has the feature of Tab History too, I tried it and it was a pleasant surprise for me, that it appeared to be way less buggy than Tab History Redux.
But the TU extension itself is a pretty huge combine, like TabMixPlus is.
So I asked the author to make a small separate extension based on TU's code but with only one feature.
He refused to do so, but he agreed to separate the code, so it can be used by extensions like userChromeJS or uc.
So you may also try to follow my way and:
1. disable Tab History Redux
2. install userChromeJS
3. use the code <a href="http://forums.mozillazine.org/viewtopic.php?f=48&t=1702465&start=1770#p11331599">from here</a>, separated from TU by it's author.

It is way less buggy, but frankly speaking I did meet the same bug a few times, but by my experience - it happens way less frequently, so it's worth switching from Tab History Redux.

**Comment by DavidRegev** [comment]
Thanks. I’ve actually been using that code for a couple of days now, and I’ve already encountered the bug where I’m unable to click the Back button. That’s when I decided to comment on this issue. Hopefully, one of these things will get fixed.

**Comment by Drugoy** [comment]
Try to find STR for that bug and report that to TU's author, please.
I don't think anyone's going to work on Tab History Redux...
